crates/driver/src/tests/cases/protocol_fees.rs

```diff
@@ -0,0 +1,143 @@
+use crate::{
+    domain::competition::order,
+    tests::{
+        self,
+        setup::{
+            ab_liquidity_quote,
+            ab_order,
+            ab_pmm_pool,
+            ab_solution,
+            ExpectedOrder,
+            FeePolicy,
+            Test,
+        },
+    },
+};
+
+async fn protocol_fee_test_case(
+    side: order::Side,
+    fee_policy: FeePolicy,
+    expected_order: ExpectedOrder,
+) {
+    let test_name = format!("Protocol Fee: {side:?} {fee_policy:?}");
+    let quote = ab_liquidity_quote()
+        .sell_amount(50000000000000000000u128.into())
+        .buy_amount(40000000000000000000u128.into());
+    let pool = ab_pmm_pool(quote);
+    let order = ab_order()
+        .kind(order::Kind::Limit)
+        .sell_amount(50000000000000000000u128.into())
+        .side(side)
+        .solver_fee(Some(10000000000000000000u128.into()))
+        .fee_policy(fee_policy)
+        .executed(40000000000000000000u128.into());
+    let test: Test = tests::setup()
+        .name(test_name)
+        .pool(pool)
+        .order(order)
+        .solution(ab_solution())
+        .done()
+        .await;
+
+    test.solve().await.ok().expected_orders(&[expected_order]);
+}
+
+#[tokio::test]
+#[ignore]
+async fn surplus_protocol_fee_buy_order_not_capped() {
+    let side = order::Side::Buy;
+    let fee_policy = FeePolicy::Surplus {
+        factor: 0.5,
+        // high enough so we don't get capped by volume fee
+        max_volume_factor: 1.0,
+    };
+    let expected = ExpectedOrder {
+        name: ab_order().name,
+        executed_sell_amount: 100000000000000000000u128.into(),
```

Comment:
Initially, the quote was part of the order. Now, it is part of the Pool since the quote affects the reserve amounts. Do we really want the quote to be part of the order? If so, different orders could potentially have different quotes while working with a single pool adjusted with a random quote?

Expected sell/buy amounts are now part of the order.

Reply:
They don't have to be part of the order, you can create a new struct `TestCase` in which you define all the variables you need for your test (limit amounts, quote amounts, executed amounts, order type etc). This way you co-locate your assumptions into a single method and don't require a poor developer that wants to understand the test 3 months from now to jump into multiple methods to get the values they need to understand the test case.

Reply:
Ah, apologies, I didn't get the initial idea correctly. Created a TestCase struct.